src/System.Management.Automation/engine/Modules/ModuleIntrinsics.cs

```diff
@@ -1208,82 +1122,126 @@ public static string GetPSModulePath(PSModulePathScope scope)
             }
         }
 
+#nullable enable
         /// <summary>
         /// Checks the various PSModulePath environment string and returns PSModulePath string as appropriate.
+        /// The PSModulePath is built in the following order:
+        /// 1. PersonalModulePath - Can be overriden by the profile config file.
+        /// 2. SharedModulePath
+        /// 3. SystemModulePath - Can be overriden by the pwsh config file.
         /// </summary>
-        public static string GetModulePath(string currentProcessModulePath, string hklmMachineModulePath, string hkcuUserModulePath)
+        /// <param name="currentProcessModulePath">The current process module path to append to the end of the env var list.</param>
+        /// <param name="hklmMachineModulePath">The system module path from the system pwsh config.</param>
+        /// <param name="hkcuUserModulePath">The personal module path from the user pwsh config.</param>
+        /// <returns>The final PSModulePath string.</returns>
+        public static string GetModulePath(string? currentProcessModulePath, string? hklmMachineModulePath, string? hkcuUserModulePath)
         {
-            string personalModulePath = GetPersonalModulePath();
-            string sharedModulePath = GetSharedModulePath(); // aka <Program Files> location
-            string psHomeModulePath = GetPSHomeModulePath(); // $PSHome\Modules location
+            string? personalModulePathToUse = string.IsNullOrEmpty(hkcuUserModulePath) ? GetPersonalModulePath() : hkcuUserModulePath;
+            string? sharedModulePath = GetSharedModulePath(); // aka <Program Files> location
+            string? systemModulePathToUse = string.IsNullOrEmpty(hklmMachineModulePath) ? GetPSHomeModulePath() : hklmMachineModulePath;
 
-            // If the variable isn't set, then set it to the default value
-            if (currentProcessModulePath == null)  // EVT.Process does Not exist - really corner case
+            List<string> finalModulePath = new List<string>();
```

Comment:
We could set initial capacity. I guess 4 is standard number of module paths (on fresh Windows) so we could use 8 to minimize allocations.

Reply:
If we really wanted to we could count the number of `Path.PathSeparator` in each of the strings and add 1 but I think settling on 8 is probably a good middle ground and not too costly to start with.

Reply:
I set it to 8 as a minimum capacity, I think that should cover the majority of scenarios.